File: btypes_primitives/src/main/java/de/hhu/stups/btypes/BTuple.java
package de.hhu.stups.btypes;

import java.util.Objects;

public final class BTuple<S, T> implements BObject {

	private final S first;
	private final T second;

	public BTuple(S first, T second) {
		this.first = Objects.requireNonNull(first, "first");
		this.second = Objects.requireNonNull(second, "second");
	}

	public boolean equals(Object o) {
		if (this == o) {
			return true;
		} else if (!(o instanceof BTuple)) {
			return false;
		} else {
			BTuple<?, ?> that = (BTuple<?, ?>) o;
			return this.first.equals(that.first) && this.second.equals(that.second);
		}
	}

	public int hashCode() {
		return Objects.hash(this.first, this.second);
	}

	@Override
	public String toString() {
		return "(" + this.first + " |-> " + this.second + ")";
	}

	public S projection1() {
		return this.first;
	}

	public T projection2() {
		return this.second;
	}

	public BBoolean equal(BTuple<?, ?> o) {
		return BBoolean.of(this.equals(o));
	}

	public BBoolean unequal(BTuple<?, ?> o) {
		return BBoolean.of(!this.equals(o));
	}
}

File: btypes_primitives/src/main/js/de/hhu/stups/btypes/BTuple.ts
import {BBoolean} from "./BBoolean";
import {BObject} from "./BObject";

export class BTuple<S,T> implements BObject {

	first: S;

	second: T;

	constructor(first: S, second: T) {
		if (first == null || second == null) {
			throw new RangeError("Arguments of Tuple cant be null or undefined");
		}
		this.first = first;
		this.second = second;
	}


	equals(o): boolean {
		if (this == o) {
			return true;
		}
		if (o == null || o !instanceof BTuple) {
			return false;
		}
		return o.projection1().equals(this.first) && o.projection2().equals(this.second);
	}

	toString(): string {
		return "(" + this.projection1() + " |-> " + this.projection2() + ')';
	}

	projection1(): S {
		return this.first;
	}

	projection2(): T {
		return this.second;
	}

	equal(o: BTuple<S, T>): BBoolean {
		return new BBoolean(this.equals(o));
	}

	unequal(o: BTuple<S, T>): BBoolean {
		return new BBoolean(!this.equals(o));
	}

}
